grub: de-hardcode firstboot network configuration

lucab opened the pull request:
This removes the hardcoded network configuration kargs from grub,
moving them into handling within initramfs by `dracut-cmdline.service`.

Ref: https://github.com/coreos/fedora-coreos-tracker/issues/460
Ref: https://github.com/coreos/fedora-coreos-config/pull/427

lucab commented:
/hold

jlebon commented:
See https://github.com/coreos/coreos-assembler/pull/1298 for the other spots we need to tweak too.

Worth noting in the commit message also that the eventual goal is to remove `rd.neednet=1` as well as we move towards on-request network enablement.

dustymabe commented:
we need to be real careful with this because it will change it for all streams. We need to make sure when we merge this and do our first testing release that we also tag coreos-assembler so we can use the same one for the subsequent stable release (this is related to https://github.com/coreos/fedora-coreos-streams/issues/64)

jlebon commented:
> we need to be real careful with this because it will change it for all streams

Yeah, good point. Tagging cosa would work, though another approach is to just add a new `image.yaml` knob which cosa keys off of. That way it naturally gets promoted together with content. I think we should do that anyway so that it's easier to ratchet all of this into FCOS/RHCOS atomically.

darkmuggle commented:
There is a potential way to deal with N streams: embed the kargs into the `ignition.firstboot` and have grub read them in and then set the defaults on a per distro/stream. The advantage here is decoupling the builder from the configuration. I also think we could take it further, and just have the `grub.cfg` provided by the stream itself and COSA place it. 

That said, I don't have a strong opinion either way. The change seems sensible to me.

jlebon commented on src/cmd-buildextend-metal:
```diff
@@ -143,6 +143,17 @@ if [ "${image_type}" = dasd ] || [ "${image_type}" = metal4k ]; then
+firstboot_kargs="$(python3 -c '
+import sys, yaml;
+LEGACY_DEFAULT = [ "rd.neednet=1", "ip=dhcp,dhcp6" ];
+args = yaml.safe_load(sys.stdin).get("extra-kargs", LEGACY_DEFAULT);
```
Hmm, I don't follow this. `extra-kargs` are persistent, but we want `ip=dhcp,dhcp6` to just be on first boot, right?

I think it might be safer/cleaner to just have a `rd-networking-by-default: true/false` type knob instead.

lucab replied:
That's actually a typo (I was planning to finish and test this PR tomorrow), it should be `firstboot-kargs`. It couples with a `image.yaml` snippet like this:
```yaml
firstboot-kargs:
    - 'rd.neednet=1'
    - 'ip=dhcp,dhcp6'
```

(None of this code has been manually tested, yet)

jlebon replied:
Ahh OK, that makes more sense. In that case, WFM. (We'll want to make sure though that we add this to the YAML files in both FCOS and RHCOS before merging.)

jlebon replied:
Hmm, thinking more on this, this still feels a bit awkward to me. Can we instead drop the `LEGACY_DEFAULT` here and just add the snippet in your comment above in `image.yaml` before merging this? Otherwise, the "pivot" point in FCOS/RHCOS would be setting `firstboot-kargs: []`, right? I think it's cleaner and clearer for the pivot to be when we remove `firstboot-kargs` entirely.

lucab replied:
Eh, we can, but this PR still has a dependency on RHCOS config plus all versioned streams.

I'd suggest we keep the `LEGACY_DEFAULT` as is for now and merge this first, then propagate the image.yaml change to all the relevant repos to make sure the defaults are in place, then possibly came back here and address the pivot point (I'm fine to handle the "no-key" case as the "empty value" case, but we should maybe leave a commented key-sample once we get there).

jlebon replied:
OK sure, WFM. I think this is the right approach overall (make default configurable, then configure it in the streams that want in, then eventually flip default here so we can drop the config knob) and we're just bikeshedding on the type of the knob. :)

See for example https://github.com/coreos/fedora-coreos-config/pull/426/files#diff-b56ad46b43eb108be045b7ad559c32b6R5. I think it would've been more obvious if the diff there were e.g.

```diff
+rd-networking-by-default: false
```

But anyway, we should be able to blow away that knob entirely soon-ish!

jlebon commented on src/grub.cfg:
```diff
@@ -44,14 +44,11 @@ fi
     # above $ignition_network_kcmdline with static networking config.
-    # This override feature is primarily used by coreos-installer to
-    # persist static networking config provided during install to the
-    # first boot of the machine.
```
Hmm, I think this comment is still true, no?

jlebon commented on src/cmd-buildextend-metal:
```diff
@@ -143,6 +143,17 @@ if [ "${image_type}" = dasd ] || [ "${image_type}" = metal4k ]; then
 
+# Firstboot kernel arguments.
+firstboot_kargs="$(python3 -c '
+import sys, yaml;
```
Minor/optional: I think we can drop the semicolons here and below since this is a proper multiline string now?

jlebon commented on src/cmd-buildextend-metal:
```diff
@@ -143,6 +143,17 @@ if [ "${image_type}" = dasd ] || [ "${image_type}" = metal4k ]; then
+import sys, yaml;
+LEGACY_DEFAULT = [ "rd.neednet=1", "ip=dhcp,dhcp6" ];
+args = yaml.safe_load(sys.stdin).get("firstboot-kargs", LEGACY_DEFAULT);
+if not args:
```
One subtlety here that I missed at first (which contributed to why I originally wanted something more explicit than this) is that it *appears* like this can never be false since we provide a default. Maybe a comment here like: `# firstboot-kargs provided, but is empty` ?

lucab replied:
(I blame PEP8 for convincing a generation of programmers that an array and a boolean are isomorphic)

lucab commented:
Rebased. There was actually a bug in the logic writing `/boot/ignition.firstboot`. I also adjusted the parameter name to match what GRUB and coreos-installer use.
https://github.com/coreos/fedora-coreos-config/pull/427 shows how to configure this logic.

This PR can land as soon review-stamped, as it internally sticks to the currently hardcoded defaults.

I manually tested that this works with the logic in Afterburn 4.4.0 in F32.

lucab commented:
/hold cancel

ashcrow commented:
> This PR can land as soon review-stamped, as it internally preserve the previously-GRUB-hardcoded defaults.

Perfect :+1: 

/lgtm

jlebon commented on src/create_disk.sh:
```diff
@@ -392,7 +395,7 @@ s390x)
 	# stage on s390x, either through zipl->grub2-emu or zipl standalone.
 	# See https://github.com/coreos/ignition-dracut/issues/84
 	# A similar hack is present in https://github.com/coreos/coreos-assembler/blob/master/src/gf-platformid#L55
-	echo "$(grep options $blsfile) ignition.firstboot rd.neednet=1 ip=dhcp,dhcp6" > $tmpfile
```
Hmm actually, I think we need to add `${firstboot_kargs}` here, right? Because s390x doesn't use the GRUB config yet. And in `gf_platformid` too.

Or... we just sprint to https://github.com/coreos/fedora-coreos-config/pull/426.

lucab replied:
This will also affect RHCOS and slower FCOS streams, so let's fix this.

I'm unsure however how to wire this new variable content to `gf_platformid`. Do we actually need to have it there, if that is only injecting the platform ID?

lucab replied:
Meh, my current attempt is at https://github.com/coreos/coreos-assembler/pull/1479.
